x/protorev/keeper/routes_test.go

```diff
@@ -0,0 +1,249 @@
+package keeper_test
+
+import (
+	"github.com/osmosis-labs/osmosis/v13/x/protorev/types"
+)
+
+type TestRoute struct {
+	PoolId      uint64
+	InputDenom  string
+	OutputDenom string
+}
+
+func (suite *KeeperTestSuite) TestBuildRoutes() {
+	cases := []struct {
+		description string
+		inputDenom  string
+		outputDenom string
+		poolID      uint64
+		expected    [][]TestRoute
+	}{
+		{
+			description: "Route exists for swap in Akash and swap out Atom",
+			inputDenom:  "akash",
+			outputDenom: types.AtomDenomination,
+			poolID:      1,
+			expected: [][]TestRoute{
+				{
+					{PoolId: 1, InputDenom: types.AtomDenomination, OutputDenom: "akash"},
+					{PoolId: 14, InputDenom: "akash", OutputDenom: "bitcoin"},
+					{PoolId: 4, InputDenom: "bitcoin", OutputDenom: types.AtomDenomination},
+				},
+				{
+					{PoolId: 25, InputDenom: types.OsmosisDenomination, OutputDenom: types.AtomDenomination},
+					{PoolId: 1, InputDenom: types.AtomDenomination, OutputDenom: "akash"},
+					{PoolId: 7, InputDenom: "akash", OutputDenom: types.OsmosisDenomination},
+				},
+			},
+		},
+		{
+			description: "Route exists for swap in Bitcoin and swap out Atom",
+			inputDenom:  "bitcoin",
+			outputDenom: types.AtomDenomination,
+			poolID:      4,
+			expected: [][]TestRoute{
+				{
+					{PoolId: 25, InputDenom: types.OsmosisDenomination, OutputDenom: types.AtomDenomination},
+					{PoolId: 4, InputDenom: types.AtomDenomination, OutputDenom: "bitcoin"},
+					{PoolId: 10, InputDenom: "bitcoin", OutputDenom: types.OsmosisDenomination},
+				},
+			},
+		},
+		{
+			description: "Route exists for swap in Bitcoin and swap out ethereum",
+			inputDenom:  "bitcoin",
+			outputDenom: "ethereum",
+			poolID:      19,
+			expected: [][]TestRoute{
```

Comment:
Asking to learn: In what case does making trade in "opposite direction" reduce arbitrage? Is there an example available?

Reply:
@davidterpay is correct.

@stackman27 if a regular user swaps OSMO -> JUNO on a single pool (pool 5 for this example), they are selling OSMO for JUNO, increasing the price of JUNO and decreasing the price of OSMO (relative prices) on pool 5. 

If you assume other OSMO/JUNO pools or routes are in sync w.r.t prices, the pool the user just swapped on now has cheap OSMO and expensive JUNO. The goal for the cyclic arbitrage is to find JUNO on a different pool that's now "cheap" compared to the JUNO on pool 5. They want to buy JUNO elsewhere, then sell JUNO on Pool 5 where it's expensive.

So when looking at pool 5, an arbitrager will always go the opposite direction as the original swap to capture the mispricing opportunity (swapping JUNO -> OSMO on pool 5).

Hope that makes sense, but ask more questions if not!